src/streamer/jamstreamer.cpp

```diff
@@ -0,0 +1,74 @@
+#ifndef _WIN32
+#include "jamstreamer.h"
+
+using namespace streamer;
+
+CJamStreamer::CJamStreamer() {}
+
+// put the received pcm data into the pipe to ffmpeg
+void CJamStreamer::process( int iServerFrameSizeSamples, const CVector<int16_t>& data ) {
+    qproc->write ( reinterpret_cast<const char*> (&data[0]), sizeof (int16_t) * ( 2 * iServerFrameSizeSamples ) );
+}
+
+void CJamStreamer::Init( const QString strStreamDest ) {
+    this->strStreamDest = strStreamDest;
+}
+
+// create a pipe to ffmpeg called "pipeout" to being able to put out the pcm data to it
+void CJamStreamer::OnStarted() {
+    if ( !qproc )
+    {
+        qproc = new QProcess;
+        QObject::connect ( qproc, &QProcess::errorOccurred, this, &CJamStreamer::onError );
+        QObject::connect ( qproc, QOverload<int, QProcess::ExitStatus>::of( &QProcess::finished ), this, &CJamStreamer::onFinished );
+        qproc->setStandardOutputFile ( QProcess::nullDevice() );
+    }
+    QStringList argumentList = { "-loglevel", "error",
+                                 "-y", "-f", "s16le",
+                                 "-ar", "48000", "-ac", "2",
+                                 "-i", "-" };
+    argumentList += strStreamDest.split( QRegExp("\\s+") );
```

Comment:
Perhaps we should use [`QProcess::splitCommand`](https://doc.qt.io/qt-5/qprocess.html#splitCommand) for this?

Reply:
> Perhaps we should use [`QProcess::splitCommand`](https://doc.qt.io/qt-5/qprocess.html#splitCommand) for this?

Absolutely! I didn't know about that. What I did just fixed the immediate issue (`--streamto "-f mp3 foo.mp3"`) but now that you mention it, it would probably break when attempting to escape things to ffmpeg (e.g. `--streamto "-f mp3 '/tmp/path with spaces'"`) or something. So my proposal makes things better, but is still not complete. I'm in favor of using `splitCommand`.

The only problem I see is that this function is only available as of Qt 5.15. As far as I can see, Jamulus is supposed to support earlier versions (5.9?). Therefore, a compromise might be a version check which uses `splitCommand` if available and falls back to the basic regexp for older versions.

Are you (@npostavs) going to continue with this? Or is @dingodoppelt?
I honestly only looked at this whole PR due to the potential security impact and I only continued investigating after the troubles with QProcess seemed stuck a bit.

Personally, I currently have no need for this feature at all so I would prefer if one of you could continue. :)

Reply:
@hoffie 
I'd really love to, but I obviously lack the necessary skills (I'm still overly happy that my first implementation did what it did but that's about what I can contribute since you guys are accomplished programmers and I still don't know what I'm doing here ;)

Reply:
Ah, didn't realize it 5.15 specific. 

> Therefore, a compromise might be a version check which uses `splitCommand` if available and falls back to the basic regexp for older versions.

The compromise sounds like a recipe for confusion to me. How necessary are the multiple arguments? E.g., `-f mp3` is redundant if you already give a `.mp3` filename anyway, as far as I've understood.

```
       -f fmt (input/output)
           Force input or output file format. The format is normally auto
           detected for input files and guessed from the file extension for
           output files, so this option is not needed in most cases.
```

Reply:
> E.g., -f mp3 is redundant if you already give a .mp3 filename anyway, as far as I've understood.

Yes, my example was hypothetical. No idea why I added that. The help text talks about "ffmpeg argument**s**", but not sure if or in what circumstances this would be needed. If the requirement could be dropped, this would be the simplest fix. @dingodoppelt?

Reply:
@hoffie 
I use it to stream to icecast and that won't work unless the format is specified. Since ffmpeg takes a lot of arguments (even multiple outputs are possible) this is very much needed so you get the full functionality that ffmpeg offers.
The recording to mp3 was just for test purposes so you wouldn't have to set up an icecast server to test the code. I wasn't aware, that the -f option is redundant for I haven't used it to record into a file (I use the awesome jamrecorder for that purpose)

Reply:
Not sure how we could move this forward.

> The compromise sounds like a recipe for confusion to me.

I generally agree. However, we are talking about edge cases I think (e.g. filenames/urls with spaces). It seems like the half-broken hack I've implemented works for @dingodoppelt's use case. If we can improve that code to be less half-broken on modern Qt versions, I think that's a benefit.

So I'd still vote for the compromise as I don't see any better alternatives?

Reply:
Hmm, the other possibilities would be for `--streamto` to take multiple arguments, a la `find -exec ... \;`. Or just require a single `--streamto` argument which is a shell script that calls `ffmpeg`?

Reply:
> the other possibilities would be for --streamto to take multiple arguments, a la find -exec ... \;

Not sure I'm following. Do you mean repeated `--streamto` arguments which would be concatenated to an argument list by Jamulus? This would delegate the escaping/parsing part to the shell, so we would get it for free. However, usability-wise it does not sound optimal to me.

> Or just require a single --streamto argument which is a shell script that calls ffmpeg?

Jamulus-wise, this would be the simplest form. @dingodoppelt Would that work for you as well?
It would have additional advantage that non-ffmpeg use cases could be enabled by this feature as well. It would have the downside that the input details (sample rate, etc.) would have to be specified in such a script.

Reply:
> The only problem I see is that this function is only available as of Qt 5.15. As far as I can see, Jamulus is supposed to support earlier versions (5.9?).

I just noticed that the Qt docs seem to say the split functions are only available as of Qt 5.14, so it might be trouble as well?

https://doc.qt.io/qt-5/qstring.html#split-6